lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,197 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {PersistentCache} = ChromeUtils.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const STORE_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  getRemoteSettings(name) {
+    return RemoteSettings(name).get();
```

Comment:
Although `RemoteSettings.get()` won't throw, it might return an empty array in certain cases (such as can't sync with the upstream, or IndexedDB failures etc), some care is needed here. Will comment separately at those places.

Reply:
Are we sure it'll only ever return an empty array or data? Part of me wants to do something like this just in case.

`return RemoteSettings(name).get() || [];` just to be sure.

Reply:
Oh wait, this is a promise that above wont work quite like that.

Reply:
Yep, see here https://searchfox.org/mozilla-central/source/services/settings/remote-settings.js#345